cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java

```diff
@@ -0,0 +1,180 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * System program management service for ensuring programs are running/stopped as expected
+ */
+public class SystemProgramManagementService extends AbstractRetryableScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemProgramManagementService.class);
+
+  private final long scheduleInterval;
+  private final CConfiguration cConf;
+  private final ProgramRuntimeService programRuntimeService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final AtomicReference<Map<ProgramId, Arguments>> programsEnabled;
+  private ScheduledExecutorService executor;
+
+  @Inject
+  SystemProgramManagementService(CConfiguration cConf, ProgramRuntimeService programRuntimeService,
+                                 ProgramLifecycleService programLifecycleService) {
+    super(RetryStrategies.fixDelay(cConf.getLong(Constants.AppFabric.PROGRAM_SCAN_INTERVAL), TimeUnit.SECONDS));
+    this.cConf = cConf;
+    this.scheduleInterval = cConf.getLong(Constants.AppFabric.PROGRAM_SCAN_INTERVAL);
+    this.programRuntimeService = programRuntimeService;
+    this.programLifecycleService = programLifecycleService;
+    this.programsEnabled = new AtomicReference<>();
+  }
+
+  /**
+   * Sets the map of programs that are currently enabled along with their runtime args.
+   * The programs that are not present in map will be stopped during the next run of the service.
+   *
+   * @param programsEnabled
+   */
+  public void setProgramsEnabled(Map<ProgramId, Arguments> programsEnabled) {
+    this.programsEnabled.set(new HashMap<>(programsEnabled));
+  }
+
+  @Override
+  protected long runTask() throws Exception {
+    if (programsEnabled.get() == null) {
+      LOG.debug("Programs to run not yet set, will be retried.");
+      return scheduleInterval;
+    }
+    reconcilePrograms();
+    return scheduleInterval;
+  }
+
+  @Override
+  protected ScheduledExecutorService executor() {
+    executor = Executors
+      .newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory(this.getClass().getSimpleName()));
+    return executor;
+  }
+
+  private void reconcilePrograms() throws Exception {
+    //take a copy
+    Map<ProgramId, Arguments> enabledProgramsMap = this.programsEnabled.get();
+    Set<ProgramId> enabledPrograms = enabledProgramsMap.keySet();
+    Set<ProgramId> programsRunning = new HashSet<>();
+    Set<ProgramRunId> programsToStop = new HashSet<>();
+    Map<ProgramId, SortedSet<ProgramRunId>> programRunsMap = new HashMap<>();
```

Comment:
I think you can simplify these couple maps and sets into this:

```java
Map<ProgramId, Arguments> programsEnabled = this.programsEnabled.get();
Set<ProgramRunId> programRunsToStop = new HashSet<>();

List<ProgramRuntimeService.RuntimeInfo> runtimeInfos = new ArrayList<>(programRuntimeService.listAll(ProgramType.values()));
// Sort by descending order of runId time.
runtimeInfos.sort((o1, o2) -> Long.compare(RunIds.getTime(o2.getController().getRunId(), TimeUnit.MILLISECONDS), 
                                           RunIds.getTime(o1.getController().getRunId(), TimeUnit.MILLISECONDS)));

for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
  ProgramId programId = runtimeInfo.getProgramId();
  if (!programId.getNamespaceId().equals(NamespaceId.SYSTEM)) {
    continue;
  }
  // Remove the enabled program if it already running.
  // If the program is not in the map, it means this particular run shouldn't be running (either not enabled or have multiple copies running)
  if (programsEnabled.remove(programId) == null) {
    programRunsToStop.add(programId.run(runtimeInfo.getController().getRunId());
  }
}

startAllPrograms(programsEnabled);
stopAllProgramRuns(programRunsToStop);
```

Reply:
Thanks, cleaned up.